implementation.md

```diff
@@ -217,41 +215,88 @@ instance, storing it as a JSON file, then using that on another instance.
+   allows creating an arbitrary number of constants, functions, and actions for
+   each.
+
+The distinction between the later two is best explained by example: for JIRA,
+we configure every JIRA connection with a separate file, so there is going to
+be exactly one _file ticket_ action definition and one _close ticket_ action
+for each configuration file; therefore, it uses a
+`FileBackedMatchedDefinitionRepository`. A `.constants` JSON file can create
+many constants for each file, so it uses a
+`FileBackedArbitraryDefinitionRepository`.
+
+#### File-Backed Repositories
+File-backed repositories are meant to be easy to implement.
+
+1. Create a class _T_ that implements `FileBackedConfiguration`.
+1. If the configuration file format is a JSON file, also extend
+   `AutoUpdatingJsonFile`.
+1. Implement the missing methods to read the configuration file and create any
+   necessary state.
+1. Implement a constructor in _T_ that takes `Path` or `Path` and `Definer`
+   depending on what kind of repository is used.
+1. Create a class _R_ that extends either
+   `FileBackedArbitraryDefinitionRepository` or
+   `FileBackedMatchedDefinitionRepository` with _T_ as the type argument.
+1. Annotate _R_ with `@MetaInfServices(DefinitionRepository.class)`.
+1. Create a constructor with no arguments. In the body, call `super`  as
+	 follows: `super(`_T_`.class, ".`_ext_`, `_T_`::new);` where _ext_ is the
+   extension for your configuration files.
+
+For an example of a matched repository, view
+[`JiraDefinitionRepository`](plugin/jira/src/main/java/ca/on/oicr/gsi/shesmu/jira/JiraDefinitionRepository.java)
+and
+[`JiraConnection`](plugin/jira/src/main/java/ca/on/oicr/gsi/shesmu/jira/JiraConnection.java)
+or
+[`SftpDefinitionRepository`](plugin/sftp/src/main/java/ca/on/oicr/gsi/shesmu/sftp/SftpDefintionRepository.java)
+and
+[`SftpServer`](plugin/sftp/src/main/java/ca/on/oicr/gsi/shesmu/sftp/SftpServer.java).
+
+#### Constants
 A constant is a value that can be generated with no input. It need not actually
 be constant (_e.g._, `now` is a constant). The Shesmu compiler will arbitrarily
 copy the value of a constant, so a constant should be side-effect free.
 
-1. Create a class that implements `ConstantSource`.
-1. Annotate this class with `@MetaInfServices`.
-1. Include a no-arguments constructor for the `ConstantSource` class.
-1. Return a stream of `Constant` objects for the constants to be created.
-
-There are a number of convenience methods to create constants for values. If
-something more complicated is desired, extend the `Constant` class and write
-arbitrary bytecode in the `load` method. The code must not change any value on
-the stack and must leave exactly one value of the correct type on the stack.
-
-The `RuntimeBinding` class can be used to create constants as method calls on
-the instance of a class.
-
-### Functions
+##### Manual Implementation
+When directly implementing the `DefinitionRepository` interface, create
```

Comment:
Perhaps indicate why you would need to implement manually. What use case do you see?